internal/catalog/catalog.go

```diff
@@ -166,16 +167,96 @@ func untar(dst, version string, tektonResources map[string]contract.TektonResour
 			// to wait until all operations have completed.
 			f.Close()
 
-			if filename != "README.md" {
-				if tektonResource.Checksum != sum {
-					fmt.Fprintf(os.Stderr, "%s checksum is different than the specified checksum in the catalog file: %s", sum, tektonResource.Checksum)
-					// FIXME: maybe handle *all* file before erroring out ?
-					return fmt.Errorf("invalid checksum for %s: %s != %s", filename, sum, tektonResource.Checksum)
-				}
-				fmt.Fprintf(os.Stderr, "✅ %s\n", tektonResource.Filename)
+					if filename != "README.md" {
+							if tektonResource.Checksum != sum {
+									fmt.Fprintf(os.Stderr, "%s checksum is different than the specified checksum in the catalog file: %s", sum, tektonResource.Checksum)
+									// FIXME: maybe handle *all* file before erroring out ?
+									return fmt.Errorf("invalid checksum for %s: %s != %s", filename, sum, tektonResource.Checksum)
+							}
+							fmt.Fprintf(os.Stderr, "✅ %s\n", tektonResource.Filename)
+					}
+
+					// Add "source" annotation to task YAML file
+					if strings.HasSuffix(target, ".yaml") {
+							if err := addSourceAnnotationToTask(target, resourcesURI); err != nil {
+									return err
+							}
+					}
 			}
+	}
+	
+}
+
+func addSourceAnnotationToTask(file, resourcesURI string) error {
+	// Add the new annotation
+	repoURL := extractRepositoryURL(resourcesURI)
+	// Open the Task YAML file
+	f, err := os.OpenFile(file, os.O_RDWR, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Create a scanner to read the file line by line
+	scanner := bufio.NewScanner(f)
+	var updatedContent []string
+
+	// Regular expression pattern to match the annotations in Task metadata
+	pattern := regexp.MustCompile(`^\s+annotations:\s*$`)
+
+	// Flag to indicate if the "tekton.dev/source" annotation is already present
+	var sourceAnnotationExists bool
+
+	// Read the file line by line
+	for scanner.Scan() {
```

Comment:
I am not sure I fully follow the flow. We use the regexp to detect than we are "in" the annotation block, I guess ? 
But we never really test if the annotation itself is there, do we ?

Reply:
I tried with a yaml that already has the annotation (in the resources.tar.gz), and I get
```
               annotations:
                 tekton.dev/pipelines.minVersion: "0.50.0"
            -    tekton.dev/source: "https://fake.host/repo/resources.tar.gz"
                 tekton.dev/categories: language
                 tekton.dev/tags: go
                 tekton.dev/displayName: "go crane image"
                 tekton.dev/platforms: "linux/amd64,linux/arm64"
            +    tekton.dev/source: "https://foo.is.bar/bat/man"
```

Reply:
> I tried with a yaml that already has the annotation (in the resources.tar.gz), and I get
> 
> ```
>                annotations:
>                  tekton.dev/pipelines.minVersion: "0.50.0"
>             -    tekton.dev/source: "https://fake.host/repo/resources.tar.gz"
>                  tekton.dev/categories: language
>                  tekton.dev/tags: go
>                  tekton.dev/displayName: "go crane image"
>                  tekton.dev/platforms: "linux/amd64,linux/arm64"
>             +    tekton.dev/source: "https://foo.is.bar/bat/man"
> ```

@Senjuti256 did you address this comment  ?